### PR TITLE
implement full text search for name/location boxes

### PR DESCRIPTION
**Pull request by Stephanie1125**
So here is some change in this pull request:
+ Edit the full_text_search.py to implement two boxes search(one for name, the other for location)
+ Edit the old structure and added filter for location field. 
+ Edited the search function inside the results.py.

Instead of sending a query_txt (string type), a query_list is sent into the search function in results.py.
query_list is a list contain [query_name, query_location]  and this is used in full_text_search.py because we are sending two query (query_name and query location), this is for distinguish between name and location queries.
if there is no location input, the query_list[1] will be ""(empty string).
and the query_name (query_list[0]) plays the role as the query_txt we have before (non-edit version: only contain the name part) 
I also edit the unit test/test_fulltext_search and server_test/person_note_tests.py and tools/all_tests are passed.

+ Here is the new ui for searching: <br>
search

### CODE REVIEW

File: app/full_text_search.py [hunk 1]
```diff
@@ -14,37 +14,20 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-import logging
 import re
-
+import logging
```
Comment (kevinalle) [inline app/full_text_search.py]: sort them alphabetically

File: app/full_text_search.py [hunk 2]
```diff
@@ -114,47 +98,74 @@ def create_romanized_query_txt(query_txt):
     return enclose_in_parenthesis(romanized_query)
 
 
-def get_person_ids_from_results(romanized_query, results_list):
+def is_query_match(query_txt, values, romanized_values):
+    """
+    To check if the query match the field value
```
Comment (kevinalle) [inline app/full_text_search.py]: Checks if the query matches the field value

File: app/full_text_search.py [hunk 3]
```diff
@@ -114,47 +98,74 @@ def create_romanized_query_txt(query_txt):
...
+    Args:
+        query_txt: Search query
+        values: field value for unidecode method
+        romanized_values: romanized field value
+    Returns:
+        Boolean
+    """
+    if not query_txt:
+        return True
```
Comment (kevinalle) [inline app/full_text_search.py]: empty matches everything?

File: app/full_text_search.py [hunk 4]
```diff
@@ -114,47 +98,74 @@ def create_romanized_query_txt(query_txt):
...
+    romanized_query_list = (script_variant.romanize_search_query(query_txt))
+
+    for name in romanized_query_list:
+        words = name.split(" ")
+        for word_index, word in enumerate(words):
+            if not re.search(word, values + " " + romanized_values, re.I):
+                break
+            if word_index == len(words) - 1:
+                return True
+    return False
```
Comment (kevinalle) [inline app/full_text_search.py]: Mm, all this function is weird. Can you explain what it does? (add a comment)

File: app/full_text_search.py [hunk 5]
```diff
@@ -114,47 +98,74 @@ def create_romanized_query_txt(query_txt):
...
+
+
+def get_person_ids_from_results(romanized_query_list, results_list):
     """
     Returns person record_id of persons
```
Comment (kevinalle) [inline app/full_text_search.py]: The explanation is useful. If it's no longer accurate, please update it, but don't remove it

File: app/full_text_search.py (hunk 1 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: ok!

File: app/full_text_search.py (hunk 2 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: ok!

File: app/full_text_search.py (hunk 3 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: yes, empty location query will match everything.
Inside the search function, it check if the name query is empty or not, if it is empty then it will show error message so this is just for empty location query.

File: app/full_text_search.py (hunk 4 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: This is for creating compiled regular expression for AND search. We want to make sure that every part of the name and location query match the name and location field. Unfortunately, there is no such symbol: "&" in the re.search so we combine the string and see if the query contain part of the field. Here is an example: 
If the user input the location query: tokyo
Inside the field there have: tokyo japan
And this is valid search and also you can get the result.
However, if user input the query: "tokyo taiwan"
This this not valid and will return None because it doesn't match the location field.
screenshot from 2016-08-30 15 02 55

For the old version of full text search, we use OR search because it only has one input box and we don't know which part of the query match the name field. (and in this cause if you input tokyo taiwan as location query it will work because the tokyo part match the location field and this will be a little bit weird and if we use the OR, the user can type random things into the search boxs and still get the result) ---> the goal is to make the location query more accurate to higher the accuracy of finding the right person.

File: app/full_text_search.py (hunk 5 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: ok!

File: app/main.py [hunk 6]
```diff
@@ -594,4 +594,4 @@ def head(self):
         self.response.clear()
 
 if __name__ == '__main__':
-    webapp.util.run_wsgi_app(webapp.WSGIApplication([('.*', Main)]))
+    webapp.util.run_wsgi_app(webapp.WSGIApplication([('.*', Main)]))
```
Comment (kevinalle) [inline app/main.py]: add newline at end of file

File: app/resources/query_form.html.template [hunk 7]
```diff
@@ -94,7 +94,7 @@
             {% endif %}
           </span>
         </div>
-      {% if config.use_family_name and not config.family_name_first %}
+       {% if config.use_family_name and not config.family_name_first %}
```
Comment (kevinalle) [inline app/resources/query_form.html.template]: remove extra space

File: app/results.py [hunk 8]
```diff
@@ -83,10 +91,10 @@ def search(self, query_txt):
         if not results:
             if config.get('enable_fulltext_search'):
                 results = full_text_search.search(self.repo,
-                                                  query_txt, MAX_RESULTS)
+                                                  query_dict, MAX_RESULTS)
             else:
                 results = indexing.search(self.repo,
-                                          TextQuery(query_txt), MAX_RESULTS)
+                                          TextQuery(query_txt), MAX_RESULTS)
```
Comment (kevinalle) [inline app/results.py]: remove space at end of line

File: tests/server_test_cases/person_note_tests.py [hunk 9]
```diff
@@ -4959,7 +4959,7 @@ def test_config_use_alternate_names(self):
         assert '_test_alternate_given' not in d.text
         assert '_test_alternate_family' not in d.text
 
-        self.go('/pakistan/results?query=_test_given+_test_family')
+        self.go('/pakistan/results?role=seek&query_name=_test_given+_test_family&query_location=')
```
Comment (kevinalle) [inline tests/server_test_cases/person_note_tests.py]: add a test with a real query_location

File: app/full_text_search.py [hunk 10]
```diff
@@ -114,47 +99,85 @@ def create_romanized_query_txt(query_txt):
...
+    Returns:
+        Boolean
+    """
+    if not query_txt:
+        return True
+
+    romanized_query_list = (script_variant.romanize_search_query(query_txt))
+
+    for name in romanized_query_list:
+        words = name.split(" ")
```
Comment (kevinalle) [inline app/full_text_search.py]: you don't need to recompute words in every loop. put this line before the for

Comment (kevinalle) [inline app/full_text_search.py]: Name this search_terms

File: app/full_text_search.py [hunk 11]
```diff
@@ -114,47 +99,85 @@ def create_romanized_query_txt(query_txt):
...
+        for word_index, word in enumerate(words):
+            if not re.search(word, values_romanized_unidecode + " " +
+                    values_romanized_jp_words, re.I):
+                break
+            if word_index == len(words) - 1:
+                return True
+    return False
```
Comment (kevinalle) [inline app/full_text_search.py]: This logic is a little hard to follow. at least add comments.
# A query matches a record if all search_terms appear in the record

File: app/main.py (hunk 6 above)
Comment (Stephanie1125) [inline app/main.py]: ok!

File: app/resources/query_form.html.template (hunk 7 above)
Comment (Stephanie1125) [inline app/resources/query_form.html.template]: ok!

File: app/results.py (hunk 8 above)
Comment (Stephanie1125) [inline app/results.py]: ok!

File: app/full_text_search.py (hunk 10 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: ok!

File: app/full_text_search.py (hunk 11 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: ok!

File: tests/server_test_cases/person_note_tests.py (hunk 9 above)
Comment (Stephanie1125) [inline tests/server_test_cases/person_note_tests.py]: ok!

File: app/full_text_search.py [hunk 12]
```diff
@@ -99,22 +99,25 @@ def create_romanized_query_txt(query_txt):
     return enclose_in_parenthesis(romanized_query)
 
 
-def is_query_match(query_txt, values_romanized_unidecode,
-                   values_romanized_jp_words):
+def search_terms(query_txt, values_romanized_unidecode,
```
Comment (kevinalle) [inline app/full_text_search.py]: ah, sorry, I meant reaname the "words" variable to search_terms. The name of this function was ok!

File: app/full_text_search.py [hunk 13]
```diff
@@ -99,22 +99,25 @@ def create_romanized_query_txt(query_txt):
...
     """
+    # empty matches everything
     if not query_txt:
         return True
 
     romanized_query_list = (script_variant.romanize_search_query(query_txt))
 
+    # A query matches a record if all search_terms appear in the record
     for name in romanized_query_list:
         words = name.split(" ")
```
Comment (kevinalle) [inline app/full_text_search.py]: rename this variable to search_terms

**Comment by kevinalle** [comment]
Cool!
Just that one comment, and LGTM

File: app/full_text_search.py (hunk 13 above)
Comment (Stephanie1125) [inline app/full_text_search.py]: Changed it to 
```
    for search_terms in romanized_query_list:
        words = search_terms.split(" ")
```

**Comment by Stephanie1125** [comment]
maybe update new message for #189 ?
